File: golabs/gocloud/godata/go-gumball-riak/gumball-riak/src/github.com/basho/taste-of-riak/csharp/Ch03-Msgy-Schema/Models/Timeline.cs
namespace Ch03_Msgy_Schema.Models
{
    using System;
    using System.Collections.Generic;

    public class Timeline : IModel
    {
        private readonly string id;
        private readonly IList<string> msgKeys = new List<string>();

        public Timeline(string id)
        {
            if (string.IsNullOrWhiteSpace(id))
            {
                throw new ArgumentNullException("id", "id must not be empty or null");
            }
            this.id = id;
        }

        public string ID
        {
            get { return id; }
        }

        public IEnumerable<string> MsgKeys
        {
            get { return msgKeys; }
        }

        public void AddMsg(string msgKey)
        {
            if (string.IsNullOrWhiteSpace(msgKey))
            {
                throw new ArgumentNullException("msgKey", "msgKey must not be null, empty or whitespace");
            }
            msgKeys.Add(msgKey);
        }

        public enum TimelineType
        {
            Inbox,
            Sent
        }
    }
}

File: golabs/gocloud/godata/go-gumball-riak/gumball-riak/src/github.com/basho/taste-of-riak/nodejs/Ch03-Msgy-Schema/models/timeline.js
'use strict';
var Model = require('./model');
var inherits = require('util').inherits;

function Timeline(id) {
    Model.call(this);

    this.data = Object.freeze({
        id: id,
        msgKeys: []
    })

    Model.defineProperties(this, this.data);
}

inherits(Timeline, Model);

Timeline.prototype.addMsg = function (msgKey) {
    this.msgKeys.push(msgKey);
};

var TimelineType = Object.freeze({ Inbox : 'INBOX', Sent : 'SENT' });

module.exports = Timeline;
module.exports.TimelineType = TimelineType;

